QuickFIXn/DataDictionary/DataDictionary.cs

```diff
@@ -574,65 +591,73 @@ private void ParseMsgEl(XmlNode node, DDMap ddmap, bool? componentRequired)
                 {
-                    if (FieldsByName.ContainsKey(fieldName) == false)
-                    {
-                        throw new Exception(
-                            $"Field '{fieldName}' is used in '{messageTypeName}', but is not defined in <fields> set.");
-                    }
+                    case "field":
+                        if (FieldsByName.ContainsKey(nameAttribute) == false)
+                        {
+                            throw new DictionaryParseException(
+                                $"Field '{nameAttribute}' is not defined in <fields> section.");
+                        }
 
-                    DDField fld = FieldsByName[fieldName];
-                    XmlAttribute req = childNode.Attributes["required"];
-                    if (req != null && req.Value == "Y"
-                        && (componentRequired == null || componentRequired.Value == true))
-                    {
-                        ddmap.ReqFields.Add(fld.Tag);
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
+                        DDField fld = FieldsByName[nameAttribute];
+                        if (required && (componentRequired == null || componentRequired.Value == true))
+                        {
+                            ddmap.ReqFields.Add(fld.Tag);
+                        }
+                        if (!ddmap.IsField(fld.Tag))
+                        {
+                            ddmap.Fields.Add(fld.Tag, fld);
+                        }
 
-                    // if first field in group, make it the DELIM
-                    if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
-                    {
-                        ((DDGrp)ddmap).Delim = fld.Tag;
-                    }
-                }
-                else if (childNode.Name == "group")
-                {
-                    DDField fld = FieldsByName[fieldName];
-                    DDGrp grp = new DDGrp();
-                    XmlAttribute req = childNode.Attributes["required"];
-                    if (req != null && req.Value == "Y"
-                        && (componentRequired == null || componentRequired.Value == true))
-                    {
-                        ddmap.ReqFields.Add(fld.Tag);
-                        grp.Required = true;
-                    }
-                    if (!ddmap.IsField(fld.Tag))
-                    {
-                        ddmap.Fields.Add(fld.Tag, fld);
-                    }
-                    grp.NumFld = fld.Tag;
-                    ParseMsgEl(childNode, grp);
-                    ddmap.Groups.Add(fld.Tag, grp);
+                        // if this is in a group whose delim is unset, then this must be the delim (i.e. first field)
+                        if ((ddmap.GetType() == typeof(DDGrp) && ((DDGrp)ddmap).Delim == 0))
```

Comment:
```suggestion
                        if (ddmap is DDGrp ddGroup && ddGroup.Delim == 0)
                        {
                            ddGroup.Delim = fld.Tag;
                        }
```

Reply:
never seen this syntax before...

Reply:
Pattern matching - I think it was introduced with C# 7 - much safer than casting.